src/mesh/mesh.cpp

```diff
@@ -1042,85 +1013,76 @@ void Mesh::ApplyUserWorkBeforeOutput(ParameterInput *pin) {
-              // send conserved variables
-      for (int i = 0; i < nmb; ++i) {
-        block_list[i]->meshblock_data.Get()->SendBoundaryBuffers();
-      }
-      call++; // 3
+    // prepare to receive conserved variables
+    for (int i = 0; i < nmb; ++i) {
+      block_list[i]->meshblock_data.Get()->StartReceiving(BoundaryCommSubset::mesh_init);
+    }
+    // send conserved variables
+    for (int i = 0; i < nmb; ++i) {
+      block_list[i]->meshblock_data.Get()->SendBoundaryBuffers();
+    }
 
-      // wait to receive conserved variables
-      for (int i = 0; i < nmb; ++i) {
-        block_list[i]->meshblock_data.Get()->ReceiveAndSetBoundariesWithWait();
-      }
-      call++; // 4
-      for (int i = 0; i < nmb; ++i) {
-        block_list[i]->meshblock_data.Get()->ClearBoundary(BoundaryCommSubset::mesh_init);
-      }
-      call++;
-      // Now do prolongation, compute primitives, apply BCs
-      for (int i = 0; i < nmb; ++i) {
-        auto &pmb = block_list[i];
-        if (multilevel) {
-          ProlongateBoundaries(pmb->meshblock_data.Get());
-        }
-        ApplyBoundaryConditions(pmb->meshblock_data.Get());
-        // Call MeshBlockData based FillDerived functions
-        Update::FillDerived(pmb->meshblock_data.Get().get());
-      }
-      const int num_partitions = DefaultNumPartitions();
-      for (int i = 0; i < num_partitions; i++) {
-        auto &md = mesh_data.GetOrAdd("base", i);
-        // Call MeshData based FillDerived functions
-        Update::FillDerived(md.get());
+    // wait to receive conserved variables
+    for (int i = 0; i < nmb; ++i) {
+      block_list[i]->meshblock_data.Get()->ReceiveAndSetBoundariesWithWait();
+    }
+    for (int i = 0; i < nmb; ++i) {
+      block_list[i]->meshblock_data.Get()->ClearBoundary(BoundaryCommSubset::mesh_init);
+    }
+    // Now do prolongation, compute primitives, apply BCs
+    for (int i = 0; i < nmb; ++i) {
+      auto &pmb = block_list[i];
+      if (multilevel) {
+        ProlongateBoundaries(pmb->meshblock_data.Get());
       }
+      ApplyBoundaryConditions(pmb->meshblock_data.Get());
+      // Call MeshBlockData based FillDerived functions
+      Update::FillDerived(pmb->meshblock_data.Get().get());
+    }
+    const int num_partitions = DefaultNumPartitions();
+    for (int i = 0; i < num_partitions; i++) {
+      auto &md = mesh_data.GetOrAdd("base", i);
+      // Call MeshData based FillDerived functions
+      Update::FillDerived(md.get());
+    }
 
-      if (!res_flag && adaptive) {
-        for (int i = 0; i < nmb; ++i) {
-          block_list[i]->pmr->CheckRefinementCondition();
-        }
+    if (init_problem && adaptive) {
+      for (int i = 0; i < nmb; ++i) {
+        block_list[i]->pmr->CheckRefinementCondition();
       }
-    } // omp parallel
+    }
 
-    if (!res_flag && adaptive) {
-      iflag = false;
-      int onb = nbtotal;
+    if (init_problem && adaptive) {
+      init_done = false;
+      const int onb = nbtotal;
```

Comment:
What does the o stand for in onb? A comment would be helpful.

Reply:
See above, I'll add the comment in the next iteration of this function.